File: Design Patterns/Prototype/prototype.py
import copy

###############################################
# Prototype Class: Character
###############################################

class Character:
    def __init__(self, name: str, health: int, attack_power: int, level: int):
        self.name = name
        self.health = health
        self.attack_power = attack_power
        self.level = level

    def clone(self):
        """
        Clone method (Prototype Pattern).
        Uses shallow copy since attributes are primitive types.
        """
        return copy.copy(self)

    def show_character_info(self):
        """
        Display character details.
        """
        print(f"Character [Name={self.name}, Health={self.health}, "
              f"AttackPower={self.attack_power}, Level={self.level}]")

###############################################
# Factory using Prototype
###############################################

class CharacterFactory:
    def __init__(self):
        # Prototype character with default attributes
        self.prototype_character = Character("DefaultName", 100, 50, 1)

    def create_character_with_new_name(self, name: str) -> Character:
        cloned = self.prototype_character.clone()
        cloned.name = name
        return cloned

    def create_character_with_new_level(self, level: int) -> Character:
        cloned = self.prototype_character.clone()
        cloned.level = level
        return cloned

    def create_character_with_new_attack_power(self, attack_power: int) -> Character:
        cloned = self.prototype_character.clone()
        cloned.attack_power = attack_power
        return cloned

###############################################
# Client Code
###############################################

if __name__ == "__main__":
    print("Prototype Pattern in Python")

    factory = CharacterFactory()

    # Create a character with a new name
    hero = factory.create_character_with_new_name("Hero")
    hero.show_character_info()

    # Create another character with a new level
    boss = factory.create_character_with_new_level(10)
    boss.show_character_info()

    # Create another character with a new attack power
    warrior = factory.create_character_with_new_attack_power(80)
    warrior.show_character_info()

File: Design Patterns/Creational/Prototype/prototype.cpp
#include <iostream>
#include <string>
using namespace std;

///////////////////////////////////////////////
// Prototype Class: Character
///////////////////////////////////////////////

class Character
{
private:
  string name;
  int health;
  int attackPower;
  int level;

public:
  // Constructor
  Character(string name, int health, int attackPower, int level)
      : name(name), health(health), attackPower(attackPower), level(level) {}

  // Clone method (Prototype Pattern)
  Character *clone() const
  {
    return new Character(*this); // Shallow copy using copy constructor
  }

  // Show character info
  void showCharacterInfo() const
  {
    cout << "Character [Name=" << name
         << ", Health=" << health
         << ", AttackPower=" << attackPower
         << ", Level=" << level
         << "]" << endl;
  }

  // Getters (needed for factory to access private members)
  string getName() const { return name; }
  int getHealth() const { return health; }
  int getAttackPower() const { return attackPower; }
  int getLevel() const { return level; }
};

///////////////////////////////////////////////
// Factory using Prototype
///////////////////////////////////////////////

class CharacterFactory
{
private:
  Character *prototypeCharacter; // Prototype object

public:
  // Constructor to create default prototype
  CharacterFactory()
  {
    prototypeCharacter = new Character("DefaultName", 100, 50, 1);
  }

  // Destructor
  ~CharacterFactory()
  {
    delete prototypeCharacter;
  }

  // Create character with new name
  Character *createCharacterWithNewName(const string &name)
  {
    Character *cloned = prototypeCharacter->clone();
    return new Character(name, cloned->getHealth(),
                         cloned->getAttackPower(), cloned->getLevel());
  }

  // Create character with new level
  Character *createCharacterWithNewLevel(int level)
  {
    Character *cloned = prototypeCharacter->clone();
    return new Character(cloned->getName(), cloned->getHealth(),
                         cloned->getAttackPower(), level);
  }

  // Create character with new attack power
  Character *createCharacterWithNewAttackPower(int attackPower)
  {
    Character *cloned = prototypeCharacter->clone();
    return new Character(cloned->getName(), cloned->getHealth(),
                         attackPower, cloned->getLevel());
  }
};

///////////////////////////////////////////////
// Main function
///////////////////////////////////////////////

int main()
{
  cout << "Prototype Pattern in C++" << endl;

  CharacterFactory factory;

  // Create a new character with a different name
  Character *hero = factory.createCharacterWithNewName("Hero");
  hero->showCharacterInfo();

  // Create another character with a different level
  Character *boss = factory.createCharacterWithNewLevel(10);
  boss->showCharacterInfo();

  // Create another character with a different attack power
  Character *warrior = factory.createCharacterWithNewAttackPower(80);
  warrior->showCharacterInfo();

  // Clean up
  delete hero;
  delete boss;
  delete warrior;

  return 0;
}
